src/llvm-alloc-opt.cpp

```diff
@@ -0,0 +1,733 @@
+        }
+    }
+}
+
+static void addRetNoAlias(Function *F)
+{
+#if JL_LLVM_VERSION >= 50000
+    F->addAttribute(AttributeList::ReturnIndex, Attribute::NoAlias);
+#else
+    F->addAttribute(AttributeSet::ReturnIndex, Attribute::NoAlias);
+#endif
+}
+
+bool AllocOpt::doInitialization(Module &M)
+{
+    ctx = &M.getContext();
+    DL = &M.getDataLayout();
+
+    alloc_obj = M.getFunction("julia.gc_alloc_obj");
+    if (!alloc_obj)
+        return false;
+
+    ptr_from_objref = M.getFunction("julia.pointer_from_objref");
+
+    T_pjlvalue = alloc_obj->getReturnType();
+    T_pjlvalue0 = PointerType::get(cast<PointerType>(T_pjlvalue)->getElementType(), 0);
+    T_pjlvalue_der = PointerType::get(cast<PointerType>(T_pjlvalue)->getElementType(),
+                                      AddressSpace::Derived);
+    T_ppjlvalue0 = PointerType::get(T_pjlvalue, 0);
+    T_ppjlvalue_der = PointerType::get(T_pjlvalue, AddressSpace::Derived);
+
+    T_int8 = Type::getInt8Ty(*ctx);
+    T_int32 = Type::getInt32Ty(*ctx);
+    T_int64 = Type::getInt64Ty(*ctx);
+    T_size = sizeof(void*) == 8 ? T_int64 : T_int32;
+    T_pint8 = PointerType::get(T_int8, 0);
+
+    if (!(pool_alloc = M.getFunction("jl_gc_pool_alloc"))) {
+        std::vector<Type*> alloc_pool_args(0);
+        alloc_pool_args.push_back(T_pint8);
+        alloc_pool_args.push_back(T_int32);
+        alloc_pool_args.push_back(T_int32);
+        pool_alloc = Function::Create(FunctionType::get(T_pjlvalue, alloc_pool_args, false),
+                                      Function::ExternalLinkage, "jl_gc_pool_alloc", &M);
+        addRetNoAlias(pool_alloc);
+    }
+    if (!(big_alloc = M.getFunction("jl_gc_big_alloc"))) {
+        std::vector<Type*> alloc_big_args(0);
+        alloc_big_args.push_back(T_pint8);
+        alloc_big_args.push_back(T_size);
+        big_alloc = Function::Create(FunctionType::get(T_pjlvalue, alloc_big_args, false),
+                                     Function::ExternalLinkage, "jl_gc_big_alloc", &M);
+        addRetNoAlias(big_alloc);
+    }
+
+#if JL_LLVM_VERSION >= 50000
+    lifetime_start = Intrinsic::getDeclaration(&M, Intrinsic::lifetime_start, { T_pint8 });
+    lifetime_end = Intrinsic::getDeclaration(&M, Intrinsic::lifetime_end, { T_pint8 });
+#else
+    lifetime_start = Intrinsic::getDeclaration(&M, Intrinsic::lifetime_start);
+    lifetime_end = Intrinsic::getDeclaration(&M, Intrinsic::lifetime_end);
+#endif
+
+    MDNode *tbaa_data;
+    MDNode *tbaa_data_scalar;
+    std::tie(tbaa_data, tbaa_data_scalar) = tbaa_make_child("jtbaa_data");
+    tbaa_tag = tbaa_make_child("jtbaa_tag", tbaa_data_scalar).first;
+
+    return true;
+}
+
+bool AllocOpt::checkInst(Instruction *I, CheckInstStack &stack, std::set<Instruction*> &uses,
+                         bool &ignore_tag)
+{
+    uses.clear();
+    if (I->user_empty())
+        return true;
+    CheckInstFrame cur{I, 0, I->user_begin(), I->user_end()};
+    stack.clear();
```

Comment:
This one might actually not be empty due to the short circuit return. (The other two should.)